cpp/src/arrow/json/parser.cc

```diff
@@ -0,0 +1,1210 @@
+        break;
+      case Type::BOOL:
+        RETURN_NOT_OK(
+            visitor.Visit(static_cast<AdaptiveBooleanBuilder*>(builder), &replacement));
+        break;
+      case Type::INT64:
+      case Type::DOUBLE:
+        RETURN_NOT_OK(
+            visitor.Visit(static_cast<Int64OrDoubleBuilder*>(builder), &replacement));
+        break;
+      case Type::TIMESTAMP:
+      case Type::STRING:
+        RETURN_NOT_OK(
+            visitor.Visit(static_cast<TimestampOrStringBuilder*>(builder), &replacement));
+        break;
+      case Type::STRUCT:
+        RETURN_NOT_OK(
+            visitor.Visit(static_cast<AdaptiveStructBuilder*>(builder), &replacement));
+        break;
+      case Type::LIST:
+        RETURN_NOT_OK(
+            visitor.Visit(static_cast<AdaptiveListBuilder*>(builder), &replacement));
+        break;
+      default:
+        ARROW_LOG(FATAL) << "How did we get here";
+        return Status::Invalid("How did we get here");
+    };
+
+    if (replacement) {
+      builder_ = replacement.get();
+      int field_index = field_index_stack_.back();
+      if (field_index == -1) {
+        auto parent = static_cast<AdaptiveListBuilder*>(builder_stack_.back());
+        parent->SetValueBuilder(std::move(replacement));
+      } else {
+        auto parent = static_cast<AdaptiveStructBuilder*>(builder_stack_.back());
+        parent->SetFieldBuilder(field_index, std::move(replacement));
+      }
+    }
+
+    return Status::OK();
+  }
+
+  AdaptiveArrayBuilder* builder_;
+  Status status_;
+  std::vector<AdaptiveArrayBuilder*> builder_stack_;
+  std::vector<std::vector<bool>> absent_fields_;
+  std::vector<int> field_index_stack_;
+};
+
+template <unsigned Flags, typename Handler>
+Status BlockParser::DoParse(Handler& handler, const char* data, uint32_t size,
+                            uint32_t* out_size) {
+  rapidjson::GenericInsituStringStream<rapidjson::UTF8<>> ss(const_cast<char*>(data));
+  rapidjson::Reader reader;
+
+  for (num_rows_ = 0; num_rows_ != max_num_rows_; ++num_rows_) {
+    // parse a single line of JSON
+    auto ok = reader.Parse<Flags>(ss, handler);
+    switch (ok.Code()) {
+      case rapidjson::kParseErrorNone:
+        // parse the next object
+        continue;
+      case rapidjson::kParseErrorDocumentEmpty: {
+        // parsed all objects, finish
+        *out_size = static_cast<uint32_t>(ss.Tell());
+        return Status::OK();
+      }
+      case rapidjson::kParseErrorTermination:
+        // handler emitted an error
+        return handler.Error();
+      default:
+        // rapidjson emitted an error
+        return ParseError() << rapidjson::GetParseError_En(ok.Code());
+    }
+  }
+  return Status::Invalid("Exceeded maximum rows");
+}
+
+Status BlockParser::Parse(const char* data, uint32_t size, uint32_t* out_size) {
+  constexpr unsigned parse_flags =
+      rapidjson::kParseInsituFlag | rapidjson::kParseIterativeFlag |
```

Comment:
AFAIK, you can't use in-situ parsing with a `const char*`, as it will modify the input.

Reply:
Oops. There's a lot of copypasta from your CSV reader. I'll correct this, thanks